File: gpt6/python/engine.py
"""Reference planner: Python standard library only. No model-supplied scores."""
from __future__ import annotations
import copy
import hashlib
import json
import math
import re
from dataclasses import dataclass
from typing import Any


def number(value: Any, label: str, lo: float = 0, hi: float = math.inf) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ValueError(f"{label}: expected a number")
    if not math.isfinite(value) or not lo <= value <= hi:
        raise ValueError(f"{label}: out of range")
    return float(value)


def text(value: Any, label: str) -> str:
    if not isinstance(value, str) or not value.strip() or len(value) > 10000:
        raise ValueError(f"{label}: expected nonempty text, max 10000 characters")
    return value


def identifier(value: Any, label: str) -> str:
    value = text(value, label)
    if not re.fullmatch(r"[A-Za-z0-9_:.-]{1,128}", value):
        raise ValueError(f"{label}: invalid identifier")
    return value


def identifiers(value: Any, label: str) -> list[str]:
    if not isinstance(value, list):
        raise ValueError(f"{label}: expected a list")
    values = [identifier(v, label) for v in value]
    if len(values) != len(set(values)):
        raise ValueError(f"{label}: duplicate identifier")
    return values


def entropy(p: float) -> float:
    number(p, "probability", 0, 1)
    return 0.0 if p in (0, 1) else -p * math.log2(p) - (1-p) * math.log2(1-p)


def information_gain(prior: float, sensitivity: float, false_positive: float) -> float:
    """I(H;Y) in bits for a binary observation, conditional on a usable result."""
    p = number(prior, "prior", 0, 1)
    s = number(sensitivity, "sensitivity", 0, 1)
    f = number(false_positive, "false_positive", 0, 1)
    q = p*s + (1-p)*f
    return max(0.0, entropy(q) - p*entropy(s) - (1-p)*entropy(f))


def posterior(prior: float, sensitivity: float, false_positive: float, positive: bool) -> float:
    p = number(prior, "prior", 0, 1)
    s = number(sensitivity, "sensitivity", 0, 1)
    f = number(false_positive, "false_positive", 0, 1)
    if type(positive) is not bool:
        raise ValueError("positive: expected boolean")
    a, b = (s, f) if positive else (1-s, 1-f)
    denominator = p*a + (1-p)*b
    if denominator == 0:
        raise ValueError("Observation impossible under the declared model; review the model")
    return p*a / denominator


@dataclass(frozen=True)
class Task:
    id: str
    title: str
    profile: str
    facts: tuple[str, ...]
    depends_on: tuple[str, ...]
    acceptance: str

    @classmethod
    def parse(cls, raw: Any) -> Task:
        keys = {"id", "title", "profile", "facts", "depends_on", "acceptance"}
        if not isinstance(raw, dict) or set(raw) != keys:
            raise ValueError("Task: incorrect fields; LLM may not supply scores or commands")
        return cls(identifier(raw["id"], "id"), text(raw["title"], "title"),
                   identifier(raw["profile"], "profile"), tuple(identifiers(raw["facts"], "facts")),
                   tuple(identifiers(raw["depends_on"], "depends_on")), text(raw["acceptance"], "acceptance"))

    def key(self, goal: str) -> str:
        # ASCII serialization is identical in both reference implementations.
        raw = json.dumps([goal, self.profile, sorted(self.facts)], ensure_ascii=True, separators=(",", ":"))
        return hashlib.sha256(raw.encode()).hexdigest()


def validate_state(state: dict) -> None:
    text(state["goal"], "goal")
    for name in ("budget", "information_weight", "min_score"):
        number(state[name], name, 0, 1e12)
    number(state["max_risk"], "max_risk", 0, 1e12)
    facts = state["facts"]
    if not isinstance(facts, list):
        raise ValueError("facts: expected a list")
    identifiers([f["id"] for f in facts], "fact IDs")
    for fact in facts:
        text(fact["text"], "fact text")
        text(fact["source"], "fact source")
        if fact["status"] not in ("observed", "reported", "disputed"):
            raise ValueError("Invalid fact status")
    for value in state["hypotheses"].values():
        number(value, "hypothesis", 0, 1)
    for profile in state["profiles"].values():
        number(profile["alpha"], "alpha", 1e-12, 1e12)
        number(profile["beta"], "beta", 1e-12, 1e12)
        for name in ("benefit", "cost", "risk"):
            number(profile[name], name, 0, 1e12)
        for name in ("sensitivity", "false_positive"):
            number(profile[name], name, 0, 1)
        if type(profile["allowed"]) is not bool:
            raise ValueError("allowed: expected boolean")
        for name in ("diagnostic_hypothesis", "effect_hypothesis"):
            if profile[name] is not None and profile[name] not in state["hypotheses"]:
                raise ValueError("Unknown profile hypothesis")
    for name in ("completed_keys", "completed_ids", "successful_ids", "seen_evidence", "seen_events"):
        identifiers(state[name], name)


def rank(state: dict, raw_tasks: Any) -> dict:
    validate_state(state)
    if state["pending"] is not None:
        raise ValueError("Observe or resolve the pending task before another plan")
    if not isinstance(raw_tasks, list) or len(raw_tasks) > 8:
        raise ValueError("Expected at most eight candidate tasks")
    tasks = [Task.parse(t) for t in raw_tasks]
    identifiers([t.id for t in tasks], "task IDs")
    known_facts = {f["id"] for f in state["facts"]}
    completed = set(state["completed_ids"])
    successful = set(state["successful_ids"])
    candidate_ids = {t.id for t in tasks}
    valid: list[dict] = []
    rejected: list[dict] = []
    keys: set[str] = set()
    # Entire proposal is rejected if any citation/profile/dependency is invented.
    for task in tasks:
        if not task.facts or not set(task.facts) <= known_facts:
            raise ValueError(f"{task.id}: missing or unknown facts")
        if task.profile not in state["profiles"]:
            raise ValueError(f"{task.id}: unknown profile")
        if not set(task.depends_on) <= completed | candidate_ids or task.id in task.depends_on:
            raise ValueError(f"{task.id}: invalid dependencies")
    # Validate the proposed dependency DAG.
    by_id = {t.id: t for t in tasks}
    visiting: set[str] = set()
    visited: set[str] = set()
    def visit(tid: str) -> None:
        if tid in visiting:
            raise ValueError("Cyclic dependencies")
        if tid in visited or tid not in by_id:
            return
        visiting.add(tid)
        for dependency in by_id[tid].depends_on:
            visit(dependency)
        visiting.remove(tid)
        visited.add(tid)
    for task in tasks:
        visit(task.id)
    for task in tasks:
        profile = state["profiles"][task.profile]
        key = task.key(state["goal"])
        reason = None
        if task.id in completed or key in state["completed_keys"] or key in keys:
            reason = "duplicate"
        elif not set(task.depends_on) <= successful:
            reason = "blocked"
        elif not profile["allowed"] or profile["risk"] > state["max_risk"]:
            reason = "policy"
        elif profile["cost"] > state["budget"]:
            reason = "budget"
        # Deduplicate feasible candidates only, so a blocked duplicate cannot hide one.
        if reason:
            rejected.append({"id": task.id, "reason": reason})
            continue
        keys.add(key)
        p = profile["alpha"] / (profile["alpha"] + profile["beta"])
        h = profile["diagnostic_hypothesis"]
        ig = 0.0 if h is None else information_gain(state["hypotheses"][h], profile["sensitivity"], profile["false_positive"])
        effect = profile["effect_hypothesis"]
        relevance = 1.0 if effect is None else state["hypotheses"][effect]
        gain = profile["benefit"] * relevance
        score = p * (gain + state["information_weight"] * ig) - profile["cost"] - profile["risk"]
        valid.append({"task": {"id": task.id, "title": task.title, "profile": task.profile,
                               "facts": list(task.facts), "depends_on": list(task.depends_on), "acceptance": task.acceptance},
                      "key": key, "probability": p, "information_gain": ig, "expected_benefit": p*gain, "score": score})
    valid.sort(key=lambda v: (-v["score"], v["task"]["id"]))
    selected = valid[0] if valid and valid[0]["score"] > state["min_score"] else None
    return {"selected": selected, "ranking": valid, "rejected": rejected,
            "status": "selected" if selected else "stop_or_request_evidence"}


def accept_plan(state: dict, decision: dict) -> dict:
    updated = copy.deepcopy(state)
    updated["pending"] = copy.deepcopy(decision["selected"])
    return updated


def observe(state: dict, event: dict, evidence: bytes) -> dict:
    validate_state(state)
    required = {"event_id", "task_id", "success", "positive", "summary"}
    if not isinstance(event, dict) or set(event) != required:
        raise ValueError("Invalid observation fields")
    identifier(event["event_id"], "event_id")
    text(event["summary"], "summary")
    if type(event["success"]) is not bool:
        raise ValueError("success must be a verified binary result, not unknown")
    if not evidence or len(evidence) > 1_000_000:
        raise ValueError("Evidence must contain 1..1000000 bytes")
    digest = hashlib.sha256(evidence).hexdigest()
    if event["event_id"] in state["seen_events"] or digest in state["seen_evidence"]:
        raise ValueError("Reused evidence/event cannot be counted twice")
    pending = state["pending"]
    if pending is None or event["task_id"] != pending["task"]["id"]:
        raise ValueError("No matching pending task")
    updated = copy.deepcopy(state)
    profile = updated["profiles"][pending["task"]["profile"]]
    h = profile["diagnostic_hypothesis"]
    if h is not None and event["success"]:
        updated["hypotheses"][h] = posterior(updated["hypotheses"][h], profile["sensitivity"], profile["false_positive"], event["positive"])
    elif event["positive"] is not None:
        raise ValueError("A failed/non-diagnostic task must have positive=null")
    profile["alpha" if event["success"] else "beta"] += 1
    updated["budget"] = max(0.0, updated["budget"] - profile["cost"])
    updated["completed_keys"].append(pending["key"])
    updated["completed_ids"].append(event["task_id"])
    if event["success"]:
        updated["successful_ids"].append(event["task_id"])
    updated["seen_evidence"].append(digest)
    updated["seen_events"].append(event["event_id"])
    fact_id = "observation:" + event["event_id"]
    if fact_id in {f["id"] for f in updated["facts"]}:
        raise ValueError("Duplicate observation fact ID")
    updated["facts"].append({"id": fact_id, "text": event["summary"], "status": "reported",
                             "source": "evidence/" + digest + ".bin"})
    # Imported observation is 'reported': bytes/hash do not prove its interpretation.
    updated["pending"] = None
    return updated

File: gpt6/typescript/engine.ts
/** Reference planner. Node built-ins only; external JSON is runtime-validated. */
import { createHash } from 'node:crypto';

export type Profile = {
  alpha: number; beta: number; benefit: number; cost: number; risk: number; allowed: boolean;
  diagnostic_hypothesis: string | null; effect_hypothesis: string | null;
  sensitivity: number; false_positive: number;
};
export type Fact = { id: string; text: string; source: string; status: 'observed' | 'reported' | 'disputed' };
export type Task = { id: string; title: string; profile: string; facts: string[]; depends_on: string[]; acceptance: string };
export type Ranked = { task: Task; key: string; probability: number; information_gain: number; expected_benefit: number; score: number };
export type State = {
  goal: string; budget: number; information_weight: number; max_risk: number; min_score: number;
  facts: Fact[]; hypotheses: Record<string, number>; profiles: Record<string, Profile>;
  completed_keys: string[]; completed_ids: string[]; successful_ids: string[]; seen_evidence: string[]; seen_events: string[];
  pending: Ranked | null;
};
export type Decision = { selected: Ranked | null; ranking: Ranked[]; rejected: {id: string; reason: string}[]; status: string };
export type Observation = {event_id: string; task_id: string; success: boolean; positive: boolean | null; summary: string};

export function object(x: unknown): Record<string, unknown> {
  if (x === null || typeof x !== 'object' || Array.isArray(x)) throw new Error('Expected object');
  return x as Record<string, unknown>;
}
export function number(x: unknown, label: string, lo = 0, hi = Infinity): number {
  if (typeof x !== 'number' || !Number.isFinite(x) || x < lo || x > hi) throw new Error(`${label}: invalid number`);
  return x;
}
export function text(x: unknown, label: string): string {
  if (typeof x !== 'string' || !x.trim() || x.length > 10000) throw new Error(`${label}: invalid text`);
  return x;
}
export function identifier(x: unknown, label: string): string {
  const value = text(x, label);
  if (!/^[A-Za-z0-9_:.-]{1,128}$/.test(value)) throw new Error(`${label}: invalid identifier`);
  return value;
}
export function identifiers(x: unknown, label: string): string[] {
  if (!Array.isArray(x)) throw new Error(`${label}: expected array`);
  const values = x.map(v => identifier(v, label));
  if (new Set(values).size !== values.length) throw new Error(`${label}: duplicate ID`);
  return values;
}
function exactKeys(x: Record<string, unknown>, keys: string[]): void {
  if (Object.keys(x).sort().join('|') !== keys.sort().join('|')) throw new Error('Incorrect fields');
}
export function entropy(p: number): number {
  number(p, 'probability', 0, 1);
  return p === 0 || p === 1 ? 0 : -p*Math.log2(p) - (1-p)*Math.log2(1-p);
}
export function informationGain(p: number, s: number, f: number): number {
  number(p, 'prior', 0, 1); number(s, 'sensitivity', 0, 1); number(f, 'false_positive', 0, 1);
  const q = p*s + (1-p)*f;
  return Math.max(0, entropy(q) - p*entropy(s) - (1-p)*entropy(f));
}
export function posterior(p: number, s: number, f: number, positive: boolean): number {
  number(p, 'prior', 0, 1); number(s, 'sensitivity', 0, 1); number(f, 'false_positive', 0, 1);
  if (typeof positive !== 'boolean') throw new Error('positive: expected boolean');
  const [a, b] = positive ? [s, f] : [1-s, 1-f];
  const denominator = p*a + (1-p)*b;
  if (denominator === 0) throw new Error('Observation impossible under the declared model; review the model');
  return p*a/denominator;
}
export function parseTask(raw: unknown): Task {
  const r = object(raw);
  exactKeys(r, ['id','title','profile','facts','depends_on','acceptance']);
  return {id:identifier(r.id,'id'), title:text(r.title,'title'), profile:identifier(r.profile,'profile'),
    facts:identifiers(r.facts,'facts'), depends_on:identifiers(r.depends_on,'depends_on'), acceptance:text(r.acceptance,'acceptance')};
}
export function taskKey(task: Task, goal: string): string {
  const ascii = JSON.stringify([goal, task.profile, [...task.facts].sort()])
    .replace(/[\u007f-\uffff]/g, c => '\\u'+c.charCodeAt(0).toString(16).padStart(4,'0'));
  return createHash('sha256').update(ascii).digest('hex');
}
export function parseState(raw: unknown): State {
  const r = object(raw);
  const factsRaw = r.facts;
  if (!Array.isArray(factsRaw)) throw new Error('facts: expected array');
  const facts: Fact[] = factsRaw.map(rawFact => {
    const f = object(rawFact);
    const status = f.status;
    if (status !== 'observed' && status !== 'reported' && status !== 'disputed') throw new Error('Invalid fact status');
    return {id:identifier(f.id,'fact id'), text:text(f.text,'fact text'), source:text(f.source,'fact source'), status};
  });
  identifiers(facts.map(f=>f.id), 'fact IDs');
  const hypotheses: Record<string,number> = Object.create(null);
  for (const [k,v] of Object.entries(object(r.hypotheses))) hypotheses[k] = number(v,'hypothesis',0,1);
  const profiles: Record<string,Profile> = Object.create(null);
  for (const [k,v] of Object.entries(object(r.profiles))) {
    const p = object(v);
    function hypothesis(field: string): string | null {
      if (p[field] === null) return null;
      const id = identifier(p[field],field);
      if (!Object.hasOwn(hypotheses,id)) throw new Error('Unknown profile hypothesis');
      return id;
    }
    if (typeof p.allowed !== 'boolean') throw new Error('allowed: expected boolean');
    profiles[k] = {alpha:number(p.alpha,'alpha',1e-12,1e12), beta:number(p.beta,'beta',1e-12,1e12),
      benefit:number(p.benefit,'benefit',0,1e12), cost:number(p.cost,'cost',0,1e12), risk:number(p.risk,'risk',0,1e12), allowed:p.allowed,
      diagnostic_hypothesis:hypothesis('diagnostic_hypothesis'), effect_hypothesis:hypothesis('effect_hypothesis'),
      sensitivity:number(p.sensitivity,'sensitivity',0,1), false_positive:number(p.false_positive,'false_positive',0,1)};
  }
  let pending: Ranked | null = null;
  if (r.pending !== null) {
    const p = object(r.pending);
    pending = {task:parseTask(p.task), key:identifier(p.key,'key'), probability:number(p.probability,'probability',0,1),
      information_gain:number(p.information_gain,'information_gain',0,1), expected_benefit:number(p.expected_benefit,'expected_benefit'),
      score:number(p.score,'score',-Infinity,Infinity)};
  }
  return {goal:text(r.goal,'goal'), budget:number(r.budget,'budget',0,1e12), information_weight:number(r.information_weight,'information_weight',0,1e12),
    max_risk:number(r.max_risk,'max_risk',0,1e12), min_score:number(r.min_score,'min_score',0,1e12), facts,hypotheses,profiles,
    completed_keys:identifiers(r.completed_keys,'completed_keys'), completed_ids:identifiers(r.completed_ids,'completed_ids'),
    successful_ids:identifiers(r.successful_ids,'successful_ids'), seen_evidence:identifiers(r.seen_evidence,'seen_evidence'),
    seen_events:identifiers(r.seen_events,'seen_events'), pending};
}
export function rank(rawState: unknown, rawTasks: unknown): Decision {
  const state = parseState(rawState);
  if (state.pending !== null) throw new Error('Observe or resolve pending task before another plan');
  if (!Array.isArray(rawTasks) || rawTasks.length > 8) throw new Error('Expected at most eight tasks');
  const tasks = rawTasks.map(parseTask);
  identifiers(tasks.map(t=>t.id),'task IDs');
  const knownFacts = new Set(state.facts.map(f=>f.id));
  const completed = new Set(state.completed_ids);
  const successful = new Set(state.successful_ids);
  const candidates = new Map(tasks.map(t=>[t.id,t]));
  const visiting = new Set<string>(), visited = new Set<string>();
  for (const t of tasks) {
    if (!t.facts.length || !t.facts.every(f=>knownFacts.has(f))) throw new Error(`${t.id}: missing or unknown facts`);
    if (!Object.hasOwn(state.profiles,t.profile)) throw new Error(`${t.id}: unknown profile`);
    if (t.depends_on.includes(t.id) || !t.depends_on.every(d=>completed.has(d)||candidates.has(d))) throw new Error('Invalid dependencies');
  }
  function visit(id: string): void {
    if (visiting.has(id)) throw new Error('Cyclic dependencies');
    if (visited.has(id) || !candidates.has(id)) return;
    visiting.add(id);
    for (const d of candidates.get(id)!.depends_on) visit(d);
    visiting.delete(id); visited.add(id);
  }
  for (const t of tasks) visit(t.id);
  const valid: Ranked[] = [], rejected: {id:string;reason:string}[] = [];
  const keys = new Set<string>();
  for (const task of tasks) {
    const profile = state.profiles[task.profile], key = taskKey(task,state.goal);
    let reason: string | null = null;
    if (completed.has(task.id) || state.completed_keys.includes(key) || keys.has(key)) reason='duplicate';
    else if (!task.depends_on.every(d=>successful.has(d))) reason='blocked';
    else if (!profile.allowed || profile.risk>state.max_risk) reason='policy';
    else if (profile.cost>state.budget) reason='budget';
    if (reason) {rejected.push({id:task.id,reason});continue;}
    keys.add(key);
    const p = profile.alpha/(profile.alpha+profile.beta);
    const h = profile.diagnostic_hypothesis;
    const ig = h === null ? 0 : informationGain(state.hypotheses[h],profile.sensitivity,profile.false_positive);
    const effect = profile.effect_hypothesis;
    const gain = profile.benefit*(effect === null ? 1 : state.hypotheses[effect]);
    const score = p*(gain+state.information_weight*ig)-profile.cost-profile.risk;
    valid.push({task,key,probability:p,information_gain:ig,expected_benefit:p*gain,score});
  }
  valid.sort((a,b)=>b.score-a.score || (a.task.id<b.task.id?-1:a.task.id>b.task.id?1:0));
  const selected = valid.length && valid[0].score>state.min_score ? valid[0] : null;
  return {selected,ranking:valid,rejected,status:selected?'selected':'stop_or_request_evidence'};
}
export function acceptPlan(state: State, decision: Decision): State {
  const updated = structuredClone(state); updated.pending=structuredClone(decision.selected); return updated;
}
export function observe(rawState: unknown, rawEvent: unknown, evidence: Uint8Array): State {
  const state = parseState(rawState), event = object(rawEvent);
  exactKeys(event,['event_id','task_id','success','positive','summary']);
  const eventId=identifier(event.event_id,'event_id'), summary=text(event.summary,'summary');
  if (typeof event.success !== 'boolean') throw new Error('success: expected verified boolean result');
  if (!evidence.length || evidence.length>1000000) throw new Error('Evidence must contain 1..1000000 bytes');
  const digest=createHash('sha256').update(evidence).digest('hex');
  if (state.seen_events.includes(eventId)||state.seen_evidence.includes(digest)) throw new Error('Reused evidence/event');
  const pending=state.pending;
  if (!pending||event.task_id!==pending.task.id) throw new Error('No matching pending task');
  const updated=structuredClone(state), profile=updated.profiles[pending.task.profile];
  const h=profile.diagnostic_hypothesis;
  if (h!==null && event.success) {
    if (typeof event.positive!=='boolean') throw new Error('positive: expected boolean');
    updated.hypotheses[h]=posterior(updated.hypotheses[h],profile.sensitivity,profile.false_positive,event.positive);
  } else if (event.positive!==null) throw new Error('Failed/non-diagnostic task needs positive=null');
  profile[event.success?'alpha':'beta']+=1;
  updated.budget=Math.max(0,updated.budget-profile.cost);
  updated.completed_keys.push(pending.key); updated.completed_ids.push(pending.task.id);
  if (event.success) updated.successful_ids.push(pending.task.id);
  updated.seen_events.push(eventId); updated.seen_evidence.push(digest);
  const factId='observation:'+eventId;
  if (updated.facts.some(f=>f.id===factId)) throw new Error('Duplicate observation fact ID');
  updated.facts.push({id:factId,text:summary,status:'reported',source:'evidence/'+digest+'.bin'});
  updated.pending=null; return updated;
}
